[WIP] [WEBSITE] Added blog "The schema evolution story of Hudi"

sbernauer opened the pull request.

pratyakshsharma commented:
Thank you for raising this. Would be better if you can attach a PDF of this blog here. Will be easier to review this. @sbernauer

sbernauer commented:
Hi @pratyakshsharma i originally wrote this blog post at https://github.com/sbernauer/hudi-blog-schema-evolution/blob/main/README.md, where you can still find it. They are identical.
I was asked to open a PR for better review :)

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+---
+title: "The schema evolution story of Hudi"
```
I would prefer to keep the title bit specific. 
"Schema evolution with DeltaStreamer using KafkaSource"

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+---
+title: "The schema evolution story of Hudi"
+excerpt: "Evolve schema to keep data up to date with business"
```
can you please fix the excerpt also based on suggested title

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+---
+
+The schema used for data exchange between services can change change rapidly with new business requirements.
+Apache Hudi is often used in combination with kafka as a Event Stream where all events are transmitted according to an Record Schema. In our case a Confluent Schema Registry is used to maintain the Schemas and the different Version every Schemas can evolve.
```
minor. ".... maintain the schema and as schema evolves, newer versions are updated in the schema registry".

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+Apache Hudi is often used in combination with kafka as a Event Stream where all events are transmitted according to an Record Schema. In our case a Confluent Schema Registry is used to maintain the Schemas and the different Version every Schemas can evolve.
+
+# What do we want to achieve?
+We have multiple instances of Deltastreamer running consuming many topics with different schemas.
```
minor. Some rephrasing and also lets add a one liner about Deltastreamer just incase someone is not aware of it. 

```
"We have multiple instances of Deltastreamer running, consuming many topics with different schemas ingesting to the same hudi table. Deltastreamer is a utility in Hudi to assist in ingesting data from multiple sources like DFS, kafka, etc into Hudi. If interested, you can read more about DeltaStreamer tool [here](https://hudi.apache.org/docs/writing_data#deltastreamer)".
```

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+
+# What do we want to achieve?
+We have multiple instances of Deltastreamer running consuming many topics with different schemas.
+Ideally every Topic should be able to evolve the schema to match new business requirements. Consumers start producing data with a new schema version and the Deltastreamer picks up the new schema and ingests the data with the new schema. For this to work we run our Deltastreamer instances with the latest schema version available from the Schema Registry to ensure that we always use the freshest sc ...
```
minor (comma) : "For this to work, we run our Deltastreamer..."

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+
+
+# What is the problem?
+The normal operation looks like this. Multiple (or a single) producers write records to the kafka topic.
```
line break at the end. 
```
...to the kafka topic. <br/>
```

nsivabalan replied:
```
.... to the kafka topic. In regular flow of events, all records are in the same schema V1 and is in sync with schema registry. "
```

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+# What is the problem?
+The normal operation looks like this. Multiple (or a single) producers write records to the kafka topic.
+![Normal operation](/assets/images/blog/hudi-schema-evolution/normal_operation.png)
+Things get complicated when a producer switches to a new Writer-Schema v2 (in this case `Producer A`). `Producer B` remains on Schema v1. E.g. a attribute `myattribute` was added to the schema, resulting in schema version v2.
```
again, lets start with a line break.
```
<br/> Things get complicated ...
```

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+The normal operation looks like this. Multiple (or a single) producers write records to the kafka topic.
+![Normal operation](/assets/images/blog/hudi-schema-evolution/normal_operation.png)
+Things get complicated when a producer switches to a new Writer-Schema v2 (in this case `Producer A`). `Producer B` remains on Schema v1. E.g. a attribute `myattribute` was added to the schema, resulting in schema version v2.
+So Deltastreamer must not only be able to handle Events that suddenly have a new Schema but also parallel operation of different Schema versions.
```
Lets reword this a bit.
```
Deltastreamer is capable of handling such schema evolution, if all incoming records were evolved and serialized with evolved schema.  But the complication is that, some records are serialized with schema version V1 and some are serialized with schema version V2. 
```

nsivabalan replied:
lets add a line break at the end before the image.

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+Things get complicated when a producer switches to a new Writer-Schema v2 (in this case `Producer A`). `Producer B` remains on Schema v1. E.g. a attribute `myattribute` was added to the schema, resulting in schema version v2.
+So Deltastreamer must not only be able to handle Events that suddenly have a new Schema but also parallel operation of different Schema versions.
+![Schema evolution](/assets/images/blog/hudi-schema-evolution/schema_evolution.png)
+The default deserializer used by Hudi `io.confluent.kafka.serializers.KafkaAvroDeserializer` uses the schema that that exact record was written with for deserialization. This causes Hudi to get records with multiple different schema from the kafka client. E.g. Event #13 has the new attribute `myattribute`, Event #14 dont has the new attribute `myattribute`. This makes things complicated and error ...
```
lets add a line break at the beginning just after image

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+Things get complicated when a producer switches to a new Writer-Schema v2 (in this case `Producer A`). `Producer B` remains on Schema v1. E.g. a attribute `myattribute` was added to the schema, resulting in schema version v2.
+So Deltastreamer must not only be able to handle Events that suddenly have a new Schema but also parallel operation of different Schema versions.
+![Schema evolution](/assets/images/blog/hudi-schema-evolution/schema_evolution.png)
+The default deserializer used by Hudi `io.confluent.kafka.serializers.KafkaAvroDeserializer` uses the schema that that exact record was written with for deserialization. This causes Hudi to get records with multiple different schema from the kafka client. E.g. Event #13 has the new attribute `myattribute`, Event #14 dont has the new attribute `myattribute`. This makes things complicated and error ...
```
minor: remove extra "that".
"the schema that that exact"

nsivabalan replied:
minor. 
```
... uses the schema that the record was serialized with"

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+![Confluent Deserializer](/assets/images/blog/hudi-schema-evolution/confluent_deserializer.png)
+
+# Solution
+We can use a custom Deserializer `KafkaAvroSchemaDeserializer` and plug it into the kafka client.
```
minor. 
```
Hudi added a new custom Deserializer `KafkaAvroSchemaDeserializer` to solve this problem of different producers producing records in different schema versions, but to use the latest schema from schema registry to deserialize all the records" .
```

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+# Solution
+We can use a custom Deserializer `KafkaAvroSchemaDeserializer` and plug it into the kafka client.
+As first step the Deserializer gets the source schema from the Hudi SchemaProvider. The SchemaProvider can get the schema for example from a Confluent Schema-Registry or a file.
+The Deserializer then reads the records from the topic with the schema the record was written. As next step it will convert all the records to the source schema from the SchemaProvider, in our case the latest schema. As a result, the kafka client will return all records with a unified schema. Hudi does not need to handle different schemas inside a single batch.
```
lets add a line break at the end before the image.

nsivabalan replied:
minor. 
```
... with a unified schema i.e. the latest schema as per schema registry."
```

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+
+# Solution
+We can use a custom Deserializer `KafkaAvroSchemaDeserializer` and plug it into the kafka client.
+As first step the Deserializer gets the source schema from the Hudi SchemaProvider. The SchemaProvider can get the schema for example from a Confluent Schema-Registry or a file.
```
can we replace "source schema" to "latest schema". Not sure if readers will understand the nitty gritty details that in schema provider we have source and target schema etc. For the context of this blog, we can assume schema provider has only one schema and so we can avoid talking about "source" schema. 
Can you please fix all such places to be consistent with this.

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+![KafkaAvroSchemaDeserializer](/assets/images/blog/hudi-schema-evolution/KafkaAvroSchemaDeserializer.png)
+
+# How to use this solution
+As of the coming release 0.9.0 the KafkaAvroSchemaDeserializer is not turned on by default, instead the normal Confluent Deseializer is used.
```
```
As of upcoming release 0.9.0, normal Confluent Deserializer is used by default. One has to explicitly set KafkaAvroSchemaDeserializer as below, in order to ensure smooth schema evolution with different producers producing records in different versions. 

`hoodie.deltastreamer.source.kafka.value.deserializer.class=org.apache.hudi.utilities.deser.KafkaAvroSchemaDeserializer`
```

nsivabalan commented on website/blog/2021-08-16-schema-evolution.md:
```diff
@@ -0,0 +1,35 @@
+
+# How to use this solution
+As of the coming release 0.9.0 the KafkaAvroSchemaDeserializer is not turned on by default, instead the normal Confluent Deseializer is used.
+You must set `hoodie.deltastreamer.source.kafka.value.deserializer.class=org.apache.hudi.utilities.deser.KafkaAvroSchemaDeserializer` to use the solution described in this blog post.
```
In the last diagram, in hudi, its written as "Residing data, multiple event V1". Guess it should be V2.

sbernauer replied:
Hi @nsivabalan, the residing data is from a time before the schema evolution happened. It is therefore stored e.g. in parquet with schema version v1 AFAIK

nsivabalan reviewed (changes requested):
some work rephrases and line breaks. 
Diagrams are pretty good :)

sbernauer commented:
Thanks for you feedback @nsivabalan, i should have addressed your feedback besides one single thread.
You can find the updated version here https://github.com/sbernauer/hudi-blog-schema-evolution/blob/main/README.md

vinothchandar reviewed (commented):
Thanks for writing this @sbernauer ! Over to @nsivabalan and others to land when ready

nsivabalan commented:
@sbernauer : CI is failing. guess break line has to be "<br/>" (slash at the end). Can you fix it and build it locally and verify that the blog renders properly. 

website/README.md should have instructions to build the site locally.

sbernauer commented:
Thanks for your feedback @pratyakshsharma! I will create a follow-up PR.

sbernauer commented on website/blog/2021-08-16-kafka-custom-deserializer.md:
```diff
@@ -0,0 +1,47 @@
+
+## What do we want to achieve?
+We have multiple instances of DeltaStreamer running, consuming many topics with different schemas ingesting to multiple Hudi tables. Deltastreamer is a utility in Hudi to assist in ingesting data from multiple sources like DFS, kafka, etc into Hudi. If interested, you can read more about DeltaStreamer tool [here](https://hudi.apache.org/docs/writing_data#deltastreamer)
+Ideally every Topic should be able to evolve the schema to match new business requirements. Consumers start producing data with a new schema version and the DeltaStreamer picks up the new schema and ingests the data with the new schema. For this to work, we run our DeltaStreamer instances with the latest schema version available from the Schema Registry to ensure that we always use the freshest s ...
```
You are right, it of course are the producers starting producing :)

sbernauer commented on website/blog/2021-08-16-kafka-custom-deserializer.md:
```diff
@@ -0,0 +1,47 @@
+## Solution
+Hudi added a new custom Deserializer `KafkaAvroSchemaDeserializer` to solve this problem of different producers producing records in different schema versions, but to use the latest schema from schema registry to deserialize all the records.<br/>
+As first step the Deserializer gets the latest schema from the Hudi SchemaProvider. The SchemaProvider can get the schema for example from a Confluent Schema-Registry or a file.
+The Deserializer then reads the records from the topic with the schema the record was written. As next step it will convert all the records to the latest schema from the SchemaProvider, in our case the latest schema. As a result, the kafka client will return all records with a unified schema i.e. the latest schema as per schema registry. Hudi does not need to handle different schemas inside a sin ...
```
Your statement is more clear, adopted the blog

sbernauer commented:
And sorry for the many grammatical errors. I'm no native English speaker

vinothchandar commented:
No worries!

pratyakshsharma commented:
> And sorry for the many grammatical errors. I'm no native English speaker

That is not a problem. Thank you so much for writing this blog! :)
